### PR TITLE
qgspostgresprovider: make unique values retrieval faster

### PR DESCRIPTION
**Pull request by ptitjano**
Postgres does not provide an "index skip scan" feature. This can make "SELECT DISTINCT" queries really slow.

This issue is fixed by replacing the "SELECT DISTINCT" query with a new query which emulates the "index skip scan" behavior. This provides the same results much faster.

### CODE REVIEW

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 1]
```diff
@@ -2056,9 +2056,18 @@ QSet<QVariant> QgsPostgresProvider::uniqueValues( int index, int limit ) const
   {
     // get the field name
     QgsField fld = field( index );
-    QString sql = QStringLiteral( "SELECT DISTINCT %1 FROM %2" )
-                  .arg( quotedIdentifier( fld.name() ),
-                        mQuery );
+    // Postgres does not provide an "index skip scan" feature.
+    // This can make SELECT DISTINCT queries really slow.
+    // This query emulates the "index skip scan" behavior to quickly retrieve unique values
+    QString sql = QStringLiteral(
```
Comment (elpaso) [inline src/providers/postgres/qgspostgresprovider.cpp:2062]:
```suggestion
    const QString sql = QStringLiteral(
```

Comment (ptitjano) [reply src/providers/postgres/qgspostgresprovider.cpp:2063]: This won't work because the `sql` is modified later.

**Comment by troopa81** [comment]
@ptitjano Like we discussed, I'm not completely sure that it would work better on every occasion. It supposedly works better with index but it's worse without index, no?

**Comment by ptitjano** [comment]
> @ptitjano Like we discussed, I'm not completely sure that it would work better on every occasion. It supposedly works better with index but it's worse without index, no?

Indeed. I have updated the PR. The "faster" is now only used when there is an index.

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 2]
```diff
@@ -2056,9 +2056,51 @@ QSet<QVariant> QgsPostgresProvider::uniqueValues( int index, int limit ) const
...
+                            " AND a.attnum=ANY(ix.indkey)"
+                            " AND ns.nspname=%1"
+                            " AND t.relname=%2"
+                            " AND a.attname=%3"
+                          ).arg(
+                            QgsPostgresConn::quotedValue( mUri.schema() ),
+                            QgsPostgresConn::quotedValue( mUri.table() ),
+                            QgsPostgresConn::quotedValue( fld.name() ) );
+    QgsPostgresResult hasIndexResult( connectionRO()->LoggedPQexec( "QgsPostgresProvider", hasIndexSql ) );
+    bool fldHasIndex = hasIndexResult.PQgetvalue( 0, 0 ).toInt() == 1;
```
Comment (troopa81) [inline src/providers/postgres/qgspostgresprovider.cpp:2082]:
```suggestion
    const bool fldHasIndex = hasIndexResult.PQgetvalue( 0, 0 ).toInt() == 1;
```

**Review by troopa81 — COMMENTED** [review]
Thanks!

The documentation says
> This is much faster when the index has many equal keys

How does it play if there is no equal keys ? Is it slower/faster or the same?

**Comment by ptitjano** [comment]
> Thanks!
> 
> The documentation says
> 
> > This is much faster when the index has many equal keys
> 
> How does it play if there is no equal keys ? Is it slower/faster or the same?

If all the values in the column are different, the query becomes slower.

**Comment by nyalldawson** [comment]
> How much for let say 10k, 100k 1M values?

+1 . Can you do a series of benchtests with/without index in a whole range of circumstances and provide the before/after timing here?

**Pull request closed without merging**